builder/remotecontext/git/gitutils.go

```diff
@@ -98,22 +98,45 @@ func getRefAndSubdir(fragment string) (ref string, subdir string) {
 
 func fetchArgs(remoteURL string, ref string) []string {
 	args := []string{"fetch", "--recurse-submodules=yes"}
-	shallow := true
 
-	if urlutil.IsURL(remoteURL) {
-		res, err := http.Head(fmt.Sprintf("%s/info/refs?service=git-upload-pack", remoteURL))
-		if err != nil || res.Header.Get("Content-Type") != "application/x-git-upload-pack-advertisement" {
-			shallow = false
-		}
-	}
-
-	if shallow {
+	if supportsShallowCheckout(remoteURL) {
 		args = append(args, "--depth", "1")
 	}
 
 	return append(args, "origin", ref)
 }
 
+func supportsShallowCheckout(remoteURL string) bool {
+	// Check if the URL is an HTTP url
+	if urlutil.IsURL(remoteURL) {
+		// Only smart-HTTP git servers support shallow clone
+
+		// URL to test for the service
+		serviceURL := fmt.Sprintf("%s/info/refs?service=git-upload-pack", remoteURL)
```

Comment:
Nit: this can just be `remoteURL + "/info/refs?service=git-upload-pack"` rather than needing `fmt.Sprintf`.